flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/StringToNumericPrimitiveCastRule.java

```diff
@@ -0,0 +1,78 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions.casting;
+
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeFamily;
+
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.STRING_DATA_TO_BYTE;
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.STRING_DATA_TO_DOUBLE;
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.STRING_DATA_TO_FLOAT;
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.STRING_DATA_TO_INT;
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.STRING_DATA_TO_LONG;
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.STRING_DATA_TO_SHORT;
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.staticCall;
+
+/**
+ * {@link LogicalTypeFamily#CHARACTER_STRING} to {@link LogicalTypeFamily#INTEGER_NUMERIC} and
+ * {@link LogicalTypeFamily#APPROXIMATE_NUMERIC} cast rule.
+ */
+class StringToNumericPrimitiveCastRule
+        extends AbstractExpressionCodeGeneratorCastRule<StringData, Number> {
+
+    static final StringToNumericPrimitiveCastRule INSTANCE = new StringToNumericPrimitiveCastRule();
+
+    private StringToNumericPrimitiveCastRule() {
+        super(
+                CastRulePredicate.builder()
+                        .input(LogicalTypeFamily.CHARACTER_STRING)
+                        .target(LogicalTypeFamily.INTEGER_NUMERIC)
+                        .target(LogicalTypeFamily.APPROXIMATE_NUMERIC)
+                        .build());
+    }
+
+    @Override
+    public String generateExpression(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        switch (targetLogicalType.getTypeRoot()) {
+            case TINYINT:
+                return staticCall(STRING_DATA_TO_BYTE(), inputTerm);
+            case SMALLINT:
+                return staticCall(STRING_DATA_TO_SHORT(), inputTerm);
+            case INTEGER:
+                return staticCall(STRING_DATA_TO_INT(), inputTerm);
+            case BIGINT:
+                return staticCall(STRING_DATA_TO_LONG(), inputTerm);
+            case FLOAT:
+                return staticCall(STRING_DATA_TO_FLOAT(), inputTerm);
+            case DOUBLE:
+                return staticCall(STRING_DATA_TO_DOUBLE(), inputTerm);
+        }
+        throw new IllegalArgumentException("This is a rule bug, contact the developers");
```

Comment:
May be it makes sense to pass `getTypeRoot` to exception to simplify debug in future...
What do you think?

Reply:
I don't know, this branch should never be took anyway given that if the type is not one of those in the branch, the rule will never be matched... I can add the type root, but i don't think it helps anyway